Add tagging subscribers on the edit page [MAILPOET-4440]

lysyjan opened the pull request:
This is the first pull request to add a new tag management feature.

**How to test**
1. Create/Edit a subscriber in the administration
2. Add/Remove tags
3. Check that tags are stored properly

Note: This PR contains a DB Migration. You need to run it before testing (e.g. deactivate and activate the plugin).

[MAILPOET-4440]

[MAILPOET-4440]: https://mailpoet.atlassian.net/browse/MAILPOET-4440?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

lysyjan committed:
Create tables for subscriber tags

[MAILPOET-4440]

lysyjan committed:
Add tags entities

[MAILPOET-4440]

lysyjan committed:
Add token field form component

[MAILPOET-4440]

lysyjan committed:
Add tokenField component to the subscribers form

[MAILPOET-4440]

lysyjan committed:
Add repository for tags

[MAILPOET-4440]

lysyjan committed:
Add loading tags into subscriber edit form

[MAILPOET-4440]

lysyjan committed:
Fix typo in const name

[MAILPOET-4440]

lysyjan committed:
Reuse NewsletterOption factory

[MAILPOET-4440]

lysyjan committed:
Reuse NewsletterOption factory

[MAILPOET-4440]

lysyjan committed:
Use data factories for generating entities

[MAILPOET-4440]

lysyjan committed:
Add SubscriberTagRepository

[MAILPOET-4440]

lysyjan committed:
Save or update subscriber's tags

[MAILPOET-4440]

lysyjan committed:
Add tag factory into tests

[MAILPOET-4440]

lysyjan committed:
Add integration test for SubscribersResponseBuilder

[MAILPOET-4440]

lysyjan committed:
Add check for saving tags

[MAILPOET-4440]

lysyjan committed:
Add saving tags into acceptance test

[MAILPOET-4440]

costasovo commented on mailpoet/assets/css/src/components-plugin/_forms.scss:
```diff
@@ -218,3 +218,7 @@ progress::-moz-progress-bar {
   }
 }
+
+.mailpoet-form-field-tags label.components-form-token-field__label {
```
I wonder why do we need to hide the label? Maybe this could use some code comment.

lysyjan replied:
I haven't found how to disable the label in the Gutenberg component, but we need to hide it due to the different form design. The label doesn't fit our form, and we would have duplicity here.

costasovo replied:
Aha. Thanks! That makes sense. Could you please add a similar comment also to the code? Maybe the component will change and then we will know why this was done and it will be safe to remove it.

costasovo commented on mailpoet/lib/AdminPages/Pages/Subscribers.php:
```diff
@@ -69,6 +69,9 @@ public function __construct(
   }
 
   public function render() {
+    // Gutenberg components styles
+    $this->wp->wpEnqueueStyle('wp-components');
+
```
I think we can't simply use styles this way. Currently, we bundle components from npm packages so when we use styles from WP we may get different results for different WP versions and also when the Gutenberg plugin is active.

Here is what happened when I the Gutenberg plugin:
<img width="1088" alt="Screenshot 2022-06-29 at 11 08 45" src="https://user-images.githubusercontent.com/1082140/176405527-ae56e1f8-198b-40be-b38b-cdd635c106dd.png">

I think we have 2 options:
1) load both wp components and CSS from WordPress
2) bundle both from npm packages

lysyjan replied:
Nice catch! I will try to solve it. Thanks for your proposals.

costasovo commented on mailpoet/lib/Subscribers/SubscriberSaveController.php:
```diff
@@ -277,4 +294,27 @@ public function updateCustomFields(array $data, SubscriberEntity $subscriber): v
+    foreach ($subscriber->getSubscriberTags() as $subscriberTag) {
+      $tag = $subscriberTag->getTag();
+      if (!$tag) {
+        continue;
```
If I'm not mistaken this happens in case the tag entity was deleted but the association is still saved in DB. Am I right?

If so. I think we could cleanup the association by calling `$subscriber->getSubscriberTags()->removeElement($subscriberTag);` instead of `continue`. What do you think?

lysyjan replied:
>If I'm not mistaken this happens in case the tag entity was deleted but the association is still saved in DB. Am I right?

I think soo.

>If so. I think we could cleanup the association by calling $subscriber->getSubscriberTags()->removeElement($subscriberTag); instead of continue.

I agree with your proposal. I will change it.

costasovo commented on mailpoet/tests/integration/Subscribers/SubscriberSaveControllerTest.php:
```diff
@@ -75,6 +80,9 @@ public function testItCanUpdateASubscriber(): void {
       'segments' => [
         $segmentOne->getId(),
       ],
+      'tags' => [
```
Just a suggestion. I think we could also test a case when we add and remove tags at the same time. We had some issues with similar cases in the past so I think it would be nice if the test protect us from such issues, which we may potentially introduce in some future refactors.

lysyjan replied:
I considered it, but we don't test this behavior on segments, so I didn't do it. If it makes sense to you, I can extend it.

costasovo reviewed (changes requested):
Great job! It is nice to see we are adding this functionality!

I've found an issue with CSS when testing with the Gutenberg plugin active and added a few suggestions.

lysyjan committed:
Remove enqueue wp-component styles

[MAILPOET-4440]

lysyjan committed:
Move wordpress/components styles to higher level

[MAILPOET-4440]

lysyjan committed:
Improve token field style compatibility

[MAILPOET-4440]

lysyjan committed:
Add comment

[MAILPOET-4440]

lysyjan committed:
Remove subscriberTag when tag doesn't exist

[MAILPOET-4440]

lysyjan committed:
Extend test case with saving tags

[MAILPOET-4440]

lysyjan commented:
@costasovo I addressed all your comments. I would like to do the interactive rebase after the successful review. Feel free to ping me with other questions.

costasovo commented on mailpoet/assets/css/src/mailpoet-admin.scss:
```diff
@@ -3,3 +3,19 @@
+// Generic
+// 3rd-party styles, can be overwritten.
+
+@import '../../../node_modules/@wordpress/interface/build-style/style';
+@import '../../../node_modules/@wordpress/editor/build-style/style';
+@import '../../../node_modules/@wordpress/components/build-style/style';
+@import '../../../node_modules/@wordpress/edit-post/build-style/style';
+@import '../../../node_modules/@wordpress/edit-post/build-style/classic';
+@import '../../../node_modules/@wordpress/block-editor/build-style/style';
+@import '../../../node_modules/@wordpress/block-library/build-style/style';
+@import '../../../node_modules/@wordpress/block-library/build-style/theme';
+@import '../../../node_modules/@wordpress/block-library/build-style/editor';
+@import '../../../node_modules/@wordpress/format-library/build-style/style';
+@import '../../../node_modules/codemirror/lib/codemirror';
+@import '../../../node_modules/codemirror/theme/neo';
```
Here we are moving all styles coming from WordPress packages to the main MailPoet admin CSS file. We use only a single component in the subscriber's form. Could we move only the @wordpress/components styles instead of moving all of them?

lysyjan committed:
Move plugin style loading into assets loader

[MAILPOET-4440]

costasovo reviewed (dismissed):
LGTM! 👍

veljkho commented:
Everything's good, except one tiny thing:
- When I try to add multiple same tags but different casing like TEst, Test, TEST, test.. it will let me enter it but when saving will save the first one TEst only. This is okay to save only one, but not sure about the behavior of letting user add multiple same tags? SS https://d.pr/i/rYIkwU @lysyjan

lysyjan commented:
@veljkho Nice catch! I think that multiple tags should be created.

veljkho reviewed (changes requested):
Requesting changes as agreed above. Thanks!

lysyjan committed:
Add check to prevent duplicities

We need to add a custom check to prevent duplicities in the token field with tags.
This step is necessary because the unique index in the database is not case-sensitive.
[MAILPOET-4352]

costasovo reviewed (approved):
LGTM
